### PR TITLE
Allow using as non-global minor mode

### PR DESCRIPTION
**Pull request by yantar92**
Currently, goggles can only be activated everywhere. However, highlighting changes is not very useful in special buffers, like elfeed or org-agenda. I would prefer to have goggles active only in text/prog-mode buffers. Would it be possible to have `goggles-global-mode` and `goggles-minor-mode` for more flexibility?

### CODE REVIEW

**Comment by minad** [comment]
Agree. I guess, this is a quick thing, since most things are already buffer local. My focus is on other packages right now, so I don't have time for this now, but I would welcome a PR if you want to see this soon.

**Comment by yantar92** [comment]
I looked into the code. It is tricky to make advices buffer-local...

**Comment by minad** [comment]
The advices will be global. But there is no problem to check if goggles-mode is enabled, before increasing the goggles--active counter. Alternatively check if goggles-mode is enabled in the after-change/post-command hook.

**Comment by yantar92** [comment]
> Alternatively check if goggles-mode is enabled in the after-change/post-command hook.

Hooks can be made buffer-local. They are not the issue.

> The advices will be global. But there is no problem to check if goggles-mode is enabled, before increasing the goggles--active counter.

It is true, however, it also means that `goggles-mode` will permanently advice functions leaving garbage even after it is disabled.

**Comment by minad** [comment]
> It is true, however, it also means that goggles-mode will permanently advice functions leaving garbage even after it is disabled.

You are right about that. But the advices are written such that they are very low overhead.

**Comment by yantar92** [comment]
> You are right about that. But the advices are written such that they are very low overhead.

Agree. Though `add-function` does have an option to advice buffer-locally. Unfortunately, this functionality is not exposed by `advice-add`. But nothing prohibits calling `add-function` directly or writing buffer-local version of `advice-add`.

**Comment by minad** [comment]
That sounds like an option too. Up to now I never had the need to look inside the advice api.

File: goggles.el [hunk 1]
```diff
@@ -158,15 +158,18 @@ and activate/deactivate them separately."
        ,@(mapcar
           (lambda (f)
             `(defun ,(intern (format "goggles--adv-%s" f)) (orig &rest args)
-               (setq goggles--active (1+ goggles--active))
-               (unwind-protect (apply orig args)
-                 (setq goggles--active (- goggles--active 1)))))
+               (when (or goggles-mode goggles-global-mode)
```
Comment (minad) [inline goggles.el:161]: Why is the when conditional needed? I wonder if this is correct. If no googles mode is active, the original function is not called.

File: goggles.el [hunk 2]
```diff
@@ -158,15 +158,18 @@ and activate/deactivate them separately."
...
+                 (setq goggles--active (1+ goggles--active))
+                 (unwind-protect (apply orig args)
+                   (setq goggles--active (- goggles--active 1))))))
           funs)
-       (defun ,name (&optional disable)
+       (defun ,name (&optional local disable)
```
Comment (minad) [inline goggles.el:166]: Could we use the order "disable local" instead?

**Comment by minad** [comment]
@yantar92 Is this ready?

**Comment by yantar92** [comment]
> Is this ready?

There are still problems with the code. I will take a look at your comments and also there are issues I found myself earlier. It will take some time though. I reopen the pull request when I think it is ready.

**Pull request closed without merging**